### PR TITLE
Added ModelProviderMenu component and reused it in PromptsendButton and MessageBase

### PR DESCRIPTION
**Pull request by hpatel292-seneca**
This is Requested from  #729

### CODE REVIEW

**Comment by humphd** [comment]
Can we close #729?

File: src/components/Menu/ModelProviderMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import useMobileBreakpoint from "../../hooks/use-mobile-breakpoint";
+
+interface ModelProviderMenuProps {
+  label?: string;
+  onItemSelect: (modelId: string) => void;
+  openOnHover?: boolean;
+  menuButtonLabel?: ReactNode;
+}
+
+const ModelProviderMenu: React.FC<ModelProviderMenuProps> = ({
```
Comment (Amnish04) [inline src/components/Menu/ModelProviderMenu.tsx:35]: I would name this `ModelSelectionMenu`. @humphd What do you think?

File: src/components/PromptForm/PromptSendButton.tsx [hunk 2]
```diff
@@ -229,48 +230,21 @@ function MobilePromptSendButton({ isLoading }: PromptSendButtonProps) {
...
-
   return (
     <ButtonGroup isAttached>
-      <Button type="submit" size="sm" isLoading={isLoading} loadingText="Sending">
+      <Button
+        type="submit"
+        size="sm"
+        isLoading={isLoading}
+        loadingText="Sending"
+        borderRightRadius="0"
```
Comment (Amnish04) [inline src/components/PromptForm/PromptSendButton.tsx:246]: We shouldn't need this inside a `ButtonGroup`. The problem you're facing is because you have an invalid `div` inside ButtonGroup that breaks the following style:
image

File: src/components/PromptForm/PromptSendButton.tsx [hunk 3]
```diff
@@ -308,89 +283,16 @@ function DesktopPromptSendButton({ isLoading }: PromptSendButtonProps) {
...
-      </Menu>
+      {
+        <Box>
+          <ModelProviderMenu
+            onItemSelect={(modelId) => {
+              const model = models.find((m) => m.id === modelId);
+              if (model) setSettings({ ...settings, model });
+            }}
+          />
+        </Box>
```
Comment (Amnish04) [inline src/components/PromptForm/PromptSendButton.tsx:294]:
```suggestion
          <ModelProviderMenu
            onItemSelect={(modelId) => {
              const model = models.find((m) => m.id === modelId);
              if (model) setSettings({ ...settings, model });
            }}
          />
```

Get rid of this `Box` to solve border issues.

File: src/components/PromptForm/PromptSendButton.tsx [hunk 4]
```diff
@@ -282,6 +256,7 @@ function DesktopPromptSendButton({ isLoading }: PromptSendButtonProps) {
           }
         >
           <Button
+            borderRightRadius="0"
```
Comment (Amnish04) [inline src/components/PromptForm/PromptSendButton.tsx:259]:
```suggestion
```

File: src/components/PromptForm/PromptSendButton.tsx [hunk 5]
```diff
@@ -59,7 +60,7 @@ function MobilePromptSendButton({ isLoading }: PromptSendButtonProps) {
 
   return (
     <ButtonGroup variant="outline" isAttached>
-      <Menu placement="top" strategy="fixed" closeOnSelect={false} offset={[-90, 0]}>
+      <Menu placement="top" strategy="absolute" closeOnSelect={false} offset={[-90, 0]}>
```
Comment (Amnish04) [inline src/components/PromptForm/PromptSendButton.tsx:63]: Why change this?

File: src/components/Menu/ModelProviderMenu.tsx (hunk 1 above)
Comment (Amnish04) [inline src/components/Menu/ModelProviderMenu.tsx:35]: I would name this `ModelSelectionMenu`. @humphd What do you think?

**Review by Amnish04 — CHANGES_REQUESTED** [review]
@hpatel292-seneca I've left some feedback. We need to replace this Menu with our custom implementation that supports SubMenus so you don't have to use an ad-hoc approach.

**Comment by tarasglek** [comment]
Thanks for moving the pull req over. I agree with Amnish re reusing code more. I suspect that might fix the only bug I found..there on first load if I click retry-with and search for "3b", the menu closes itself....on subsequent uses it seems ok

File: src/components/PromptForm/PromptSendButton.tsx (hunk 5 above)
Comment (hpatel292-seneca) [reply src/components/PromptForm/PromptSendButton.tsx:63]: Initially, I changed this because previously I was using `60vh` for `MenuList` and it was overflowing in mobile devices to I tried using `absolute` if that can fix my issue but I didn't so I change the height to `50vh` and it was working fine but I forget to change it back to `fixed`.

File: src/components/PromptForm/PromptSendButton.tsx (hunk 3 above)
Comment (hpatel292-seneca) [reply src/components/PromptForm/PromptSendButton.tsx:294]: Understood.

File: src/components/Menu/ModelProviderMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+          cursor="pointer"
+          color="inherit"
+          padding="0"
+          background="none"
+          border="none"
+          aria-label={label || "Choose Model"}
+          title={label || "Choose Model"}
+          onClick={adjustedOpenOnHover ? undefined : onOpen} // Open on click if not openOnHover
+          onMouseEnter={adjustedOpenOnHover ? onOpen : undefined} // Open on hover if openOnHover
+          onMouseLeave={adjustedOpenOnHover ? onClose : undefined} // Close on hover out if openOnHover
```
Comment (Amnish04) [inline src/components/Menu/ModelProviderMenu.tsx:88]: We shouldn't need to do this ourselves once we migrate to using React-Menu.

File: src/components/Menu/ModelProviderMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          fontSize="1.25rem"
+          aria-label={label || "Choose Model"}
+          title={label || "Choose Model"}
+          icon={<TbChevronUp />}
+          onClick={openOnHover ? undefined : onOpen} // Open on click if not openOnHover
+          onMouseEnter={openOnHover ? onOpen : undefined} // Open on hover if openOnHover
+          onMouseLeave={openOnHover ? onClose : undefined} // Close on hover out if openOnHover
+          borderLeftRadius="0"
+        />
+      )}
```
Comment (Amnish04) [inline src/components/Menu/ModelProviderMenu.tsx:105]: @hpatel292-seneca Right now, this implementation is too specific to be reused anywhere as you'll have to add another `MenuButton` **inside** the component implementation wherever this needs to be used.

There are 2 ways we can solve this:
1. Either accept a `variant: "menu" | "submenu"` and a `MenuButton` in the props so you can dynamically provide the host element for menu. But this will be too dirty, and not much better than the current approach.
2. I would say a **better way** for this would be to just extract the `MenuList` part as a separate component so you could reuse it anywhere - wrap it in a `Menu` with whatever `MenuButton` or throw it inside a `SubMenu`. The component could be renamed to `ModelSelectionMenuList`. Just make sure you also import the `MenuGroup` from the library we are wrapping, i.e. [@szhsin/react-menu](https://szhsin.github.io/react-menu/docs#menu-group)

@humphd What are your thoughts on this?

**Review by Amnish04 — CHANGES_REQUESTED** [review]
@hpatel292-seneca Just wrote some implementation ideas.

File: src/components/Menu/ModelProviderMenu.tsx (hunk 7 above)
Comment (humphd) [reply src/components/Menu/ModelProviderMenu.tsx:105]: Yeah, your "better way" sounds good

**Comment by tarasglek** [comment]
> Thanks for moving the pull req over. I agree with Amnish re reusing code more. I suspect that might fix the only bug I found..there on first load if I click retry-with and search for "3b", the menu closes itself....on subsequent uses it seems ok

Sweet, functionality-wise this is perfect!

**Comment by hpatel292-seneca** [comment]
Thanks @tarasglek, I found few bugs on mobile view while working it. I will open issues for those and if you agree to make those changes, I would to work on those.

**Review by Amnish04 — APPROVED** [review]
@hpatel292-seneca This works well and code looks good to me. Thanks for following up!

**Comment by humphd** [comment]
@hpatel292-seneca can you "Resolve" the issues above that are now dealt with?  It's hard to follow the current state of this PR otherwise.

Let me know when this is ready to land.

**Comment by hpatel292-seneca** [comment]
Hi @humphd,

I can send a PR for the above issue if needed. However, if we merge the PR for the above issue first, this PR will likely encounter merge conflicts, as it also modifies the same file. Let me know how you'd like to proceed.

**Comment by humphd** [comment]
I want to get this merged first then you can improve things in follow-ups.  Please resolve whatever is done above, and file a new issue with the things to be fixed next, then resolve, so we can land this.

File: src/components/Menu/ModelProviderMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  InputLeftElement,
+  Menu,
+  MenuButton,
+  MenuDivider,
+  MenuGroup,
+  MenuItem,
+  MenuList,
+  IconButton,
+  useDisclosure,
+} from "@chakra-ui/react";
```
[the remark this answers is not in the document]
Comment (hpatel292-seneca) [reply src/components/Menu/ModelProviderMenu.tsx]: This is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/8e706c19f003c517e2bcc61b0293f65b449b416d

File: src/components/Menu/ModelProviderMenu.tsx (hunk 1 above)
Comment (hpatel292-seneca) [reply src/components/Menu/ModelProviderMenu.tsx]: I followed your suggested in below comments and named the new component `[ModelSelectionMenuList]` in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/8e706c19f003c517e2bcc61b0293f65b449b416d

File: src/components/PromptForm/PromptSendButton.tsx (hunk 2 above)
Comment (hpatel292-seneca) [reply src/components/PromptForm/PromptSendButton.tsx]: This is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/8e706c19f003c517e2bcc61b0293f65b449b416d

File: src/components/PromptForm/PromptSendButton.tsx (hunk 3 above)
Comment (hpatel292-seneca) [reply src/components/PromptForm/PromptSendButton.tsx]: This is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/5a286eb40b8abf796c9e9558a1a20320668ffc0e

File: src/components/PromptForm/PromptSendButton.tsx (hunk 4 above)
Comment (hpatel292-seneca) [reply src/components/PromptForm/PromptSendButton.tsx]: It is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/5a286eb40b8abf796c9e9558a1a20320668ffc0e

File: src/components/Menu/ModelProviderMenu.tsx (hunk 6 above)
Comment (hpatel292-seneca) [reply src/components/Menu/ModelProviderMenu.tsx]: This is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/8e706c19f003c517e2bcc61b0293f65b449b416d

File: src/components/Menu/ModelProviderMenu.tsx (hunk 7 above)
Comment (hpatel292-seneca) [reply src/components/Menu/ModelProviderMenu.tsx]: This is fixed in commit https://github.com/tarasglek/chatcraft.org/pull/730/commits/8e706c19f003c517e2bcc61b0293f65b449b416d

**Comment by hpatel292-seneca** [comment]
I resolved all 'requested changes.' Sorry, I misinterpreted your message at first.

**Comment by tarasglek** [comment]
@hpatel292-seneca this is a freaking huge quality of life improvement. Awesome work!

**Comment by hpatel292-seneca** [comment]
Thanks @tarasglek
